File: Chapitre4/exercice4.3.py
import os
import sys

def main():
    # Création du tube (pipe)
    try:
        r, w = os.pipe()
    except OSError:
        print("Erreur : pipe")
        sys.exit(1)

    # Création du processus fils
    try:
        pid = os.fork()
    except OSError:
        print("Erreur : fork")
        sys.exit(1)

    if pid == 0:
        # --- PROCESSUS FILS ---
        
        # Convertir les descripteurs de fichiers en chaînes pour les passer en arguments
        lecture = str(r)
        ecriture = str(w)

        # Préparation des arguments pour execv
        # En Python, le premier argument de la liste args doit être le nom du programme
        args = ["./fils", lecture, ecriture]

        try:
            # os.execv remplace le processus actuel par le programme spécifié
            os.execv("./fils", args)
        except OSError:
            print("Erreur : execv")
            sys.exit(1)

    else:
        # --- PROCESSUS PÈRE ---
        
        # Le père ne lit pas, on ferme le côté lecture
        os.close(r)

        message = "Bonjour tout le monde: ceci est le message"
        
        # En Python, on doit encoder la chaîne en bytes avant de l'écrire dans un pipe
        # +1 pour simuler le '\0' de fin de chaîne en C (facultatif en pur Python mais fidèle au C)
        os.write(w, (message + '\0').encode())

        # Fermer le côté écriture après l'envoi
        os.close(w)

        # Attendre la fin du fils
        os.wait()

if __name__ == "__main__":
    main()

File: Chapitre4/exercice4.3.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <string.h>

int main()
{
    int tube[2];
    if(pipe(tube) == -1){ perror("pipe"); exit(1);}

    pid_t fils = fork();
    if(fils == -1) { perror("fork"); exit(1);}

    if(fils == 0)
    {
        // Convertir les numéros en texte pour argv
        char lecture[10], ecriture[10];
        sprintf(lecture,  "%d", tube[0]);  // 3 → "3"
        sprintf(ecriture, "%d", tube[1]);  // 4 → "4"

        char *args[] = {"./fils", lecture, ecriture, NULL};
        execv("./fils", args);

        perror("execv"); 
        exit(1);
    }
    else
    {
        close(tube[0]); //le père ne lit pas
        char *message="Bonjour tout le monde: ceci est le message";
        write(tube[1], message ,strlen(message)+1);
        close(tube[1]);

        wait(NULL);

    }
    return 0;
}